src/metabase/api/database.clj

```diff
@@ -185,24 +185,47 @@
       (and (source-query-cards-exist? :card) virtual-db-metadata) (concat [virtual-db-metadata]))))
 
 (defn- filter-databases-by-data-model-perms
+  "Filters the provided list of databases by data model perms, returning only the databases for which the current user
+  can fully or partially edit the data model. If the user also has block permissions for any databases, returns only the
+  name and ID of these databases, removing all other fields."
   [dbs]
-  (if-let [f (u/ignore-exceptions
-              (classloader/require 'metabase-enterprise.advanced-permissions.common)
-              (resolve 'metabase-enterprise.advanced-permissions.common/filter-databases-by-data-model-perms))]
-    (f dbs)
-    dbs))
+  (let [filtered-dbs
+        (if-let [f (u/ignore-exceptions
+                    (classloader/require 'metabase-enterprise.advanced-permissions.common)
+                    (resolve 'metabase-enterprise.advanced-permissions.common/filter-databases-by-data-model-perms))]
+          (f dbs)
+          dbs)]
+    (map
+     (fn [db] (if (mi/can-read? db)
+                db
+                (select-keys db [:id :name :tables])))
+     filtered-dbs)))
+
+(defn- check-db-data-model-perms
+  [db include-editable-data-model?]
+  (if include-editable-data-model?
+    (let [filtered-dbs (filter-databases-by-data-model-perms [db])]
+      (when (seq filtered-dbs)
+        (first filtered-dbs)))
+    db))
 
 (defn- dbs-list [& {:keys [include-tables?
                            include-saved-questions-db?
                            include-saved-questions-tables?
-                           exclude-uneditable-data-model?
+                           include-editable-data-model?
                            exclude-uneditable-details?]}]
-  (let [dbs (filter mi/can-read? (db/select Database {:order-by [:%lower.name :%lower.engine]}))
+  (let [dbs (db/select Database {:order-by [:%lower.name :%lower.engine]})
+        dbs (if include-editable-data-model?
+              ;; Since users can have *data model* perms even if they have no *data* perms for a DB, we don't filter by
+              ;; read permissions here if `include-editable-data-model?` is true. Instead, we only return ID, name
+              ;; and tables for DBs for which the user has no data perms, to avoid exposing unecessary DB metadata.
+              dbs
+              (filter mi/can-read? dbs))
```

Comment:
Not saying you should change it as part of this PR since it already works this way but in the future it's definitely better to be filtering stuff out as part of the DB query itself as opposed to fetching and then filtering with `can-read?`... if we wanted to page this endpoint in the future it's much easier to do if we don't fetch stuff we don't need